resolver/resolver.cc

```diff
@@ -2894,23 +2896,28 @@ ast::ParsedFilesOrCancelled Resolver::resolveSigs(core::GlobalState &gs, vector<
             if (result.gotItem()) {
                 combinedTrees.insert(combinedTrees.end(), make_move_iterator(threadResult.trees.begin()),
                                      make_move_iterator(threadResult.trees.end()));
-                combinedSignatures.emplace_back(move(threadResult.signatureJobs));
-                combinedMultiSignatures.emplace_back(move(threadResult.multiSignatureJobs));
+                combinedFileJobs.insert(combinedFileJobs.end(), make_move_iterator(threadResult.fileSigs.begin()),
+                                        make_move_iterator(threadResult.fileSigs.end()));
             }
         }
     }
 
+    // We need to define sigs in a stable order since, when there are conflicting sigs in multiple RBI files, the last
+    // sig 'wins'.
```

Comment:
One thing I've always wondered was whether we should do some heuristics to determine the "best" location, i.e., maybe we should pick the sig with the fewest counts of `T.untyped` in the sig.

It seems like having this comment here is a good start in case we want to revisit that—we'd have a concrete place to look.

Reply:
iirc sorting over all sigs in a big project can still get pretty expensive; we sort them in batches by file here to make it cheaper.

I initially looked to see if we could make this process idempotent, but we don't track that we previously registered a sig on a particular method.

Hmm... now that I think of it, we could just maintain a vector that acts as a map from method symbol to sig info to determine if a sig we encounter should "override" what was seen before. So maybe there's a cheap path forward for this. This solution may also let us remove the sorting step entirely.

Reply:
even in a world where we'd have some other heuristic, we'd want a consistent way to break ties:

```ruby
sig {params(x: Integer).returns(T.untyped)}
def foo(x); end

sig {params(x: String).returns(T.untyped)}
def foo(x); end
```

Reply:
Some of the flakyness in our codebase was originating from exactly this. People seem to have the assumption that they can "refine" the signature of a method by adding an(other) RBI file. So some of the manual RBI files written like that were racing with the autogenerated DSL RBI files wrt the sigs of some of the methods.

All that to say that it would be great to have a way of pulling this off in a structured way. Breaking ties could still be "last one seen" or something like that. As long as that ordering is documented, we could redirect people to put their "override" signatures at certain locations in the file system.

Reply:
It's not documented (maybe we should do so), but right now files are ordered by their paths (the sort occurs in `options.cc`) and the "last sig wins".

Reply:
FWIW even with this fix I'm not getting signature documentation if there's more than one signature.

e.g. I have Faker sigs at `sorbet/rbi/gems/faker@2.15.1.rbi` which is generated by Tapioca and has no docs or types. I also have `sorbet/rbi/sorbet-typed/lib/faker/~>2.15/faker.rbi`, and that's generated by Sord so it does have docs and type signatures.

If I hover over `Faker::Date.between(from: 1.month.ago.to_date, to: 1.day.ago.to_date)` I don't get the documentation for it, only the type signatures. I've tried restarting VS Code and some other stuff. I assume if I moved the sorbet-typed directory around that might fix it? For now I've just chosen to delete the Tapioca-generated RBI for now since all the methods are covered by the one in sorbet-typed anyway.

Reply:
@connorshea that sounds like an unrelated issue.

This pull request was designed to ensure that sorbet returned the same typecheck results no matter in which order arbitrary typechecking threads were scheduled by the operating system. It picked **an** order, and made that order apply all the time. It made no claims to say it was the best order.

Feel free to open another feature request for the documentation on hover.